### PR TITLE
Ratio experiment

### PR DESCRIPTION
**Pull request by amnona**
Add the RatioExperiment class
It includes:
* from_exp() to get ratios of features from an existing experiment
* get_sign_pvals() to get features enriched in positive/negative ratios
* nice heatmap

Also update analysis.py for nicer code and sorting first by p-value and then by effect size (so similar effect size will be ordered by p-value) as well as a bug fix for q-values not being sorted

### CODE REVIEW

File: calour/ratio_experiment.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+class RatioExperiment(Experiment):
+    '''This class stores log-ratio data and corresponding analysis methods
+
+    This is a child class of :class:`.Experiment`.
+
+    Parameters
+    ----------
+    data : numpy.ndarray or scipy.sparse.csr_matrix
+        The abundance table for OTUs or ASVs. Samples
```
Comment (RNAer) [inline calour/ratio_experiment.py:45]: this is ratio? pls correct the `Attributes` doc too

File: calour/_testing.py [hunk 2]
```diff
@@ -62,6 +62,9 @@ def setUp(self):
         self.q2_cfs_map = join(test_data_dir, 'cfs-map.txt')
         self.q2_cfs_repseqs = join(test_data_dir, 'cfs-rep-seqs.qza')
         self.q2_cfs_taxonomy = join(test_data_dir, 'cfs-taxonomy.qza')
+        # an experiment used to create a RatioExperiment
```
Comment (RNAer) [inline calour/_testing.py:65]: these are only used in ratio_exp? if so, they should be in the setUp() in test_ratio_experiment.py. Putting them here slows down all the tests.

File: calour/ratio_experiment.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        threshold: float or None, optional
+            If not None, assign each data value<threshold to threshold. If both nominator and denominator are < threshold,
+            the resulting ratio will assigned be np.nan.
+            For example, in an AmpliconExperiment, ratios calculated between low read numbers are not reliable. So it is advised to set threshold to approx. 10
+
+        Returns
+        -------
+        RatioExperiment
+        The samples are all the unique values of common_field
+        The data values for each feature are the ratio of mean data values in samples of value1 in group_field / data values in samples of value2 in group_field
```
Comment (RNAer) [inline calour/ratio_experiment.py:143]: follow the format!!

File: calour/ratio_experiment.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        exp: calour.Experiment
+            The experiment to take the 2 groups of samples from.
+        common_field: str
+            Name of the sample metadata field to calculate the ratio within.
+            For example, to calculate the ratio between 2 timepoints for each individual,
+            common_field with be 'subject_id'
+        group_field: str
+            Name of the sample metadata field on which to divide to 2 groups for ratio calculation.
+            For example, to calculate the ratio between 2 timepoints for each individual,
+            group_field with be 'time_point'
```
Comment (RNAer) [inline calour/ratio_experiment.py:128]: let's give a better example: before treatment and after treatment?

File: calour/ratio_experiment.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        value1 : str or iterable of str
+            Values for field that will be assigned to the first (nominator) sample group
```
Comment (RNAer) [inline calour/ratio_experiment.py:130]: before treatment

File: calour/ratio_experiment.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        value2: str or iterable of str or None, optional
```
Comment (RNAer) [inline calour/ratio_experiment.py:131]: after treatment

File: calour/ratio_experiment.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            If not None, values for field that will be assigned to the second (denominator) sample group.
+            If None, use all samples not in the first sample group as the second sample group.
+        threshold: float or None, optional
+            If not None, assign each data value<threshold to threshold. If both nominator and denominator are < threshold,
+            the resulting ratio will assigned be np.nan.
+            For example, in an AmpliconExperiment, ratios calculated between low read numbers are not reliable. So it is advised to set threshold to approx. 10
+
```
Comment (RNAer) [inline calour/ratio_experiment.py:138]: what happens if there is 0 or > 1 samples for one subject before (or after) treatment? need explanation

File: calour/ratio_experiment.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+                    cmetadata.at[ccol] = 'NA (multiple values)'
+            cmetadata.at[group_field] = new_field_val
+            new_sample_metadata = new_sample_metadata.append(cmetadata)
+            found_indices.append(idx)
+
+        # keep only samples that were actually added to the ratio_mat
+        ratio_mat = ratio_mat[found_indices, :]
+        ratio_exp = RatioExperiment(data=ratio_mat, sample_metadata=new_sample_metadata, feature_metadata=exp.feature_metadata,
+                                    sparse=False, databases=exp.databases, description=exp.description, info=exp.info)
+        return ratio_exp
```
Comment (RNAer) [inline calour/ratio_experiment.py:196]: this implementation feels complicated. can we filter before treatment samples and after treatment samples? and sort them the same by subject, and finally divide.

File: calour/ratio_experiment.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            For example, to calculate the ratio between 2 timepoints for each individual,
+            group_field with be 'time_point'
+        value1 : str or iterable of str
+            Values for field that will be assigned to the first (nominator) sample group
+        value2: str or iterable of str or None, optional
+            If not None, values for field that will be assigned to the second (denominator) sample group.
+            If None, use all samples not in the first sample group as the second sample group.
+        threshold: float or None, optional
+            If not None, assign each data value<threshold to threshold. If both nominator and denominator are < threshold,
+            the resulting ratio will assigned be np.nan.
```
Comment (RNAer) [inline calour/ratio_experiment.py:136]: is `nan` a good idea in data table? will it cause problems in other data manipulation functions? I haven't checked this - just brought it up.

File: calour/ratio_experiment.py (hunk 1 above)
Comment (amnona) [reply calour/ratio_experiment.py:45]: fixed

File: calour/ratio_experiment.py (hunk 3 above)
Comment (amnona) [reply calour/ratio_experiment.py:148]: fixed

File: calour/ratio_experiment.py (hunk 4 above)
Comment (amnona) [reply calour/ratio_experiment.py:128]: fixed

File: calour/ratio_experiment.py (hunk 5 above)
Comment (amnona) [reply calour/ratio_experiment.py:130]: fixed

File: calour/ratio_experiment.py (hunk 6 above)
Comment (amnona) [reply calour/ratio_experiment.py:131]: fixed

File: calour/ratio_experiment.py (hunk 7 above)
Comment (amnona) [reply calour/ratio_experiment.py:143]: added explanation in the docstring for the function.

File: calour/ratio_experiment.py (hunk 8 above)
Comment (amnona) [reply calour/ratio_experiment.py:215]: don't see how to do it simpler. can you give more details?

File: calour/ratio_experiment.py (hunk 9 above)
Comment (amnona) [reply calour/ratio_experiment.py:136]: will check other functions to see they handle nan well.
nan is important since it means we do not know the ratio since the feature was not in the subset of samples...

File: calour/_testing.py (hunk 2 above)
Comment (amnona) [reply calour/_testing.py]: fixed

**Comment by amnona** [comment]
@RNAer ready for merge?

File: calour/ratio_experiment.py (hunk 5 above)
Comment (maque4004) [inline calour/ratio_experiment.py]: pls explain what if value1 have > 1 samples

File: calour/ratio_experiment.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+            For example, in an AmpliconExperiment, ratios calculated between low read numbers are not reliable. So it is advised to set threshold to approx. 10
+
+        Returns
+        -------
+        RatioExperiment
+        The samples are all the unique values of common_field
```
Comment (maque4004) [inline calour/ratio_experiment.py]: I think it's not all unique values in common_field, only unique values which contain value1 and value2 in group_field?

File: calour/tests/test_ratio_experiment.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+from calour.ratio_experiment import RatioExperiment
+
+
+class ExperimentTests(Tests):
+    def setUp(self):
+        super().setUp()
+        self.test1 = ca.read_amplicon(self.rat1_biom, self.rat1_samp,
+                                      min_reads=10, normalize=None)
+
+    def test_from_exp(self):
```
Comment (maque4004) [inline calour/tests/test_ratio_experiment.py:25]: suggest a more complicated example:  contain > 1 samples in value1 or value2

File: calour/tests/test_ratio_experiment.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        self.assertEqual(rexp['S3', 'AG'], np.log2(600 / 100))
+        self.assertTrue(np.isnan(rexp['S1', 'AC']))
+
+        # supply threshold
+        rexp = RatioExperiment.from_exp(self.test1, 'subj', 'time', '1', '2', threshold=400)
+        # we should lose S1 since both are < threshold
+        self.assertTrue(np.isnan(rexp['S1', 'AA']))
+        # and D3 300 is corrected to 400
+        self.assertEqual(rexp['S3', 'AA'], np.log2(400 / 500))
+
```
Comment (maque4004) [inline calour/tests/test_ratio_experiment.py:46]: need to test what if value2 is None?

File: calour/ratio_experiment.py (hunk 5 above)
Comment (amnona) [reply calour/ratio_experiment.py]: Added explanation. Hope it's clear enough :)

File: calour/ratio_experiment.py (hunk 10 above)
Comment (amnona) [reply calour/ratio_experiment.py]: cool. fixed.

File: calour/tests/test_ratio_experiment.py (hunk 11 above)
Comment (amnona) [reply calour/tests/test_ratio_experiment.py:25]: cool. added

File: calour/tests/test_ratio_experiment.py (hunk 12 above)
Comment (amnona) [reply calour/tests/test_ratio_experiment.py:46]: Cool. added test.

**Comment by amnona** [comment]
ready for merge? :)
